docs: rewrite introduction to dependency injection (#44466)
angular/angular: aio/content/guide/hierarchical-dependency-injection.md

# Hierarchical injectors

> The applications you build with Angular can become quite large, and one way to manage this complexity is to split up the application into many small well-encapsulated modules, that are by themselves split up into a well-defined tree of components.

Reviewer: ```suggestion
The applications you build with Angular can become quite large, and one way to manage this complexity is to split up the application into many small well-encapsulated modules. These modules are, in turn, split up into a well-defined tree of components.
```

> There can be sections of your page that works in a completely independent way than the rest of the application, with its own local copies of the services and other dependencies that it needs. Some of the services that these sections of the application use might be shared with other parts of the application, or with parent components that are further up in the component tree, while other dependencies are meant to be private.

Reviewer: ```suggestion
There can be sections of your page that work in a completely independent way than the rest of the application; these sections have their own local copies of the services and other dependencies that it needs. Some of the services that these sections of the application use might be shared with other parts of the application, or with parent components that are further up in the component tree, while other dependencies are meant to be private.
```

> With hierarchical dependency injection, you can isolate sections of the application and give them their own private dependencies not shared with the rest of the application, or have parent components share certain dependencies with its child components only but not with the rest of the component tree, and so on. Hierarchical dependency injection enables you to share dependencies between different parts of the application only when and if you need to.

Reviewer: ```suggestion
With hierarchical dependency injection, you can isolate sections of the application and give them their own private dependencies which are not shared with the rest of the application. These sections will also have parent components that will share certain dependencies with its child components only, but not with the rest of the component tree. Hierarchical dependency injection enables you to share dependencies between different parts of the application only when and if you need to.
```

<code-example header="providers-viewproviders/src/app/child.component.ts" path="providers-viewproviders/src/app/child/child.component.1.ts" region="flowerservice"></code-example>

Now that the `FlowerService` is provided in the `@Component()` decorator, when the `<app-child>` requests the service, the injector has only to look as far as the `ElementInjector` in the `<app-child>`.

> It b't have to continue the search any further through the injector tree.

Reviewer 2: Looks like an unneeded change?

Author: Fixed.
